text/0000-mir.md

```diff
@@ -0,0 +1,821 @@
+
+The primary kind of statement is an assignent:
+
+    LVALUE "=" RVALUE
+    
+The semantics of this operation are to first evaluate the RVALUE and
+then store it into the LVALUE (which must represent a memory location
+of suitable type). 
+
+An `LVALUE` represents a path to a memory location. This is the basic
+"unit" analyzed by the borrow checker. It is always possible to
+evaluate an `LVALUE` without triggering any side-effects (modulo
+derefences of unsafe pointers, which naturally can trigger arbitrary
+behavior if the pointer is not valid).
+
+    LVALUE = B                   // reference to a user-declared binding
+           | TEMP                // a temporary introduced by the compiler
+           | ARG                 // a formal argument of the fn
+           | STATIC              // a reference to a static or static mut
+           | RETURN              // the return pointer of the fn
+           | LVALUE.f            // project a field or tuple field, like x.f or x.0
+           | *LVALUE             // dereference a pointer
+           | LVALUE[LVALUE]      // index into an array (see disc. below about bounds checks)
+           | (LVALUE as VARIANT) // downcast to a specific variant of an enum,
+                                 // see the section on desugaring matches below
+           
+An `RVALUE` represents a computation that yields a result. This result
+must be stored in memory somewhere to be accessible. The MIR does not
+contain any kind of nested expressions: everything is flattened out,
+going through lvalues as intermediaries.
+
+    RVALUE = Use(LVALUE)                // just read an lvalue
+           | [LVALUE; LVALUE]
+           | &'REGION LVALUE
+           | &'REGION mut LVALUE
+           | LVALUE as TYPE
+           | LVALUE <BINOP> LVALUE
+           | <UNOP> LVALUE
+           | Struct { f: LVALUE0, ... } // aggregates, see section below
+           | (LVALUE...LVALUE)
+           | [LVALUE...LVALUE]
+           | CONSTANT
+           | LEN(LVALUE)                // load length from a slice, see section below
+           | BOX                        // malloc for builtin box, see section below
+    BINOP = + | - | * | / | ...         // excluding && and ||
+    UNOP = ! | -                        // note: no `*`, as that is part of LVALUE
+    
+One thing worth pointing out is that the binary and unary operators
+are only the *builtin* form, operating on scalar values. Overloaded
+operators will be desugared to trait calls. Moreover, all method calls
+are desugared into normal calls via UFCS form.
+
+### Constants
+
+Constants are a subset of rvalues that can be evaluated at compilation
+time:
+
+    CONSTANT = INT
+             | UINT
+             | FLOAT
+             | BOOL
+             | BYTES
+             | STATIC_STRING
+             | ITEM<SUBSTS>                 // reference to an item or constant etc
+             | <P0 as TRAIT<P1...Pn>>       // projection
```

Comment:
That's UFCS I assume (`<T0 as TRAIT<T1...Tn>>::item` or `<T as Inherent(DefId)>::item`).